src/appdistribution/client.ts

```diff
@@ -0,0 +1,154 @@
+import * as _ from "lodash";
+import * as api from "../api";
+import * as utils from "../utils";
+import { Distribution } from "./distribution";
+import { FirebaseError } from "../error";
+
+// tslint:disable-next-line:no-var-requires
+const pkg = require("../../package.json");
+
+/**
+ * Proxies HTTPS requests to the App Distribution server backend.
+ */
+export class AppDistributionClient {
+  static MAX_POLLING_RETRIES = 30;
+  static POLLING_INTERVAL_MS = 1000;
+
+  constructor(private readonly appId: string) {
+    this.appId = appId;
+  }
+
+  async provisionApp(): Promise<void> {
+    await api.request("POST", `/v1alpha/apps/${this.appId}`, {
+      origin: api.appDistributionOrigin,
+      auth: true,
+    });
+
+    utils.logSuccess("provisioned for app distribution");
+  }
+
+  async getJwtToken(): Promise<string> {
+    const apiResponse = await api.request("GET", `/v1alpha/apps/${this.appId}/jwt`, {
+      auth: true,
+      origin: api.appDistributionOrigin,
+    });
+
+    return _.get(apiResponse, "body.token");
+  }
+
+  async uploadDistribution(token: string, distribution: Distribution): Promise<string> {
+    const apiResponse = await api.request("POST", "/spi/v1/jwt_distributions", {
+      origin: api.appDistributionUploadOrigin,
+      headers: {
+        Authorization: `Bearer ${token}`,
+        "X-APP-DISTRO-API-CLIENT-ID": pkg.name,
+        "X-APP-DISTRO-API-CLIENT-TYPE": distribution.platform(),
+        "X-APP-DISTRO-API-CLIENT-VERSION": pkg.version,
+      },
+      files: {
+        file: {
+          stream: distribution.readStream(),
+          size: distribution.fileSize(),
+          contentType: "multipart/form-data",
+        },
+      },
+    });
+
+    return _.get(apiResponse, "response.headers.etag");
+  }
+
+  async pollReleaseIdByHash(hash: string, retryCount = 0): Promise<any> {
```

Comment:
Does this return `any`? Or does it return `string`? (Of course, in a promise, but I think this can be changed to a `string` - might need to tweak a line or two to get the typing right, but this _does_ end up returning a string...).

Reply:
It's actually a recursive function :) Updated the type to be `Promise<string> | string`.

Reply:
It's updated to be `Promise<Promise<string> | string>`, which is more confusing.

I just checked: this can be changed to `Promise<string>` with no other changes.

My understanding is because of how async/await works, the wrapping causes this confusing type to be (basically) reduced to `Promise<string>`. I'd give that a functional test as well, but I believe it to be true.

Reply:
Interesting. Updated!